periphery/src/Router.sol

```diff
@@ -105,29 +67,55 @@ contract Router {
         shares = lender.deposit(amount, msg.sender);
     }
 
-    function repayWithApprove(Lender lender, uint256 amount, address beneficiary) external returns (uint256 units) {
-        lender.asset().safeTransferFrom(msg.sender, address(lender), amount);
+    function repayWithPermit2(
+        Lender lender,
+        uint256 amount,
+        address beneficiary,
+        uint256 nonce,
+        uint256 deadline,
+        bytes calldata signature
+    ) external returns (uint256 units) {
+        // Transfer tokens from the caller to the lender.
+        PERMIT2.permitTransferFrom(
+            // The permit message.
+            IPermit2.PermitTransferFrom({
+                permitted: IPermit2.TokenPermissions({token: lender.asset(), amount: amount}),
+                nonce: nonce,
+                deadline: deadline
+            }),
+            // The transfer recipient and amount.
+            IPermit2.SignatureTransferDetails({to: address(lender), requestedAmount: amount}),
+            // The owner of the tokens, which must also be
+            // the signer of the message, otherwise this call
+            // will fail.
+            msg.sender,
+            // The packed signature that was the result of signing
+            // the EIP712 hash of `permit`.
+            signature
+        );
+
         units = lender.repay(amount, beneficiary);
     }
 
-    function repayWithPermit(
+    function isMaxRedeemDynamic(Lender lender, address owner) external view returns (bool) {
+        // NOTE: If the first statement is true, the second statement will also be true (unless this is the block in which
+        // they deposited for the first time). We include the first statement only to reduce computation.
+        return lender.courierOf(owner) > 0 || lender.balanceOf(owner) != lender.maxRedeem(owner);
+    }
+
+    function redeemWithChecks(
         Lender lender,
-        uint256 amount,
-        address beneficiary,
-        uint256 allowance,
+        uint256 shares,
         uint256 deadline,
         uint8 v,
         bytes32 r,
         bytes32 s
-    ) external returns (uint256 units) {
-        ERC20 asset = lender.asset();
+    ) external returns (uint256 amount) {
+        uint256 maxRedeem = lender.maxRedeem(msg.sender);
+        if (shares > maxRedeem) shares = maxRedeem;
 
-        if (allowance != 0) {
-            asset.permit(msg.sender, address(this), allowance, deadline, v, r, s);
-        }
-
-        asset.safeTransferFrom(msg.sender, address(lender), amount);
-        units = lender.repay(amount, beneficiary);
+        lender.permit(msg.sender, address(this), shares, deadline, v, r, s);
```

Comment:
This looks like you're using `Permit` from [EIP-2612](https://github.com/dragonfly-xyz/useful-solidity-patterns/tree/main/patterns/permit2#permit-eip-2612-model) instead of `Permit2`. Is that correct? If so, the link provided mentions that this isn't supported by all major tokens:

> Since EIP-2612 is an extension of the ERC20 standard, this functionality is only possible on new (or upgradeable) tokens. So there are very few major tokens in the wild where this pattern actually works.

Should this also be using `Permit2`?

Reply:
Yes, this is using EIP2612. The `Router` needs permission to burn the user's shares. Since that permission is governed by our `Lender` contract (which implements ERC20 and the EIP2612 extension) we don't need to worry about compatibility.